### PR TITLE
Rescue unknown checkout errors

### LINKED ISSUE #4696: No error is displayed to the user when the checkout process fails on the server side
## Description
The user is not notified when the checkout button fails on the server.

## Expected Behavior
The user should be notified that an error occurred. If no details are known a generic error message can be shown telling the user to contact support.

## Actual Behaviour
The checkout page just remains available to be edited.

## Steps to Reproduce
You can follow for example #4695

## Workaround
no workaround

## Severity
bug-s2 - this is a disaster UX problem, whenever there's a problem like #4695 the users have no clue what happened after clicking the crucial checkout button. A farmer reported "users thought the order was placed".

## Possible Fix
We need to check the checkout request to the server and print a generic error if it blows up.

### PR DESCRIPTION
**Pull request by mkllnk**
#### What? Why?

Closes #4696

The user was not notified when the checkout button fails on the server. I now catch every possible standard error, send it to Bugsnag and return a generic message to the user.

#### What should we test?

Go through the description of #4695 to trigger a failing checkout. You should see a red error message popping up.

Screenshot from 2020-01-22 10-42-39

#### Release notes

Display an error message if the checkout fails due to an unknown bug.

Changelog Category: Added

### CODE REVIEW

File: app/controllers/checkout_controller.rb [hunk 1]
```diff
@@ -68,6 +68,13 @@ def update
         render json: { path: order_path(@order) }, status: :ok
       end
     end
+  rescue Spree::Core::GatewayError => error
+    # This is done for all actions in the Spree::CheckoutController.
+    rescue_from_spree_gateway_error(error)
```
Comment (luisramos0) [inline app/controllers/checkout_controller.rb]: nice catch :+1:

File: config/locales/en.yml [hunk 2]
```diff
@@ -1181,6 +1181,7 @@ en:
     already_ordered:
       cart: "cart"
       message_html: "You have an order for this order cycle already. Check the %{cart} to see the items you ordered before. You can also cancel items as long as the order cycle is open."
+    failed: "The checkout failed. Please let us know so that we can process your order."
```
Comment (luisramos0) [inline config/locales/en.yml]: I wonder if this is the best message. From a user perspective: who is "us" and do I contact this "us"?

Comment (mkllnk) [reply config/locales/en.yml]: I know, it's ambiguous. Notifying the shop owner is probably the most important thing to sort out the order. We can see the error in Bugsnag but another prompt with details would be nice as well. It's difficult to give specific instructions if we don't know what the error is.

In the context of the shop, I hope that "us" is seen as the distributor and that people click on the Contact tab. I will ask our instance managers what they think.

Comment (sauloperez) [reply config/locales/en.yml]: in the meantime, I'm pretty sure they know the distributor and how to reach him out (facebook, twitter, website, phone number, etc.). I'd go with this and improve it later if we need to.

**Comment by RachL** [comment]
This is working great!

image

**Pull request merged**